From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: IBM Codestream linux-2.6.16 october2005, patch base-21
Patch-Mainline: 2.6.17-rc1

[patch] ipv6 shared card id and delete route objects.

IPv6 changes:
 - Copy ipv6 shared card id from real device to vlan device.
 - Delete route objects related to netdevices on NETDEV_UNREGISTER events.
Acked-by: Hannes Reinecke <hare@suse.de>

---

 net/8021q/vlan.c |    3 +++
 net/ipv6/ndisc.c |    3 +++
 2 files changed, 6 insertions(+)

diff -urpN linux-2.6/net/8021q/vlan.c linux-2.6-patched/net/8021q/vlan.c
--- linux-2.6/net/8021q/vlan.c	2006-03-20 06:53:29.000000000 +0100
+++ linux-2.6-patched/net/8021q/vlan.c	2006-03-22 11:47:03.000000000 +0100
@@ -450,6 +450,9 @@ static struct net_device *register_vlan_
 	new_dev->flags = real_dev->flags;
 	new_dev->flags &= ~IFF_UP;
 
+	/* ipv6 shared card related stuff */
+	new_dev->dev_id = real_dev->dev_id;
+
 	new_dev->state = real_dev->state & VLAN_LINK_STATE_MASK;
 
 	/* need 4 bytes for extra VLAN header info,
diff -urpN linux-2.6/net/ipv6/ndisc.c linux-2.6-patched/net/ipv6/ndisc.c
--- linux-2.6/net/ipv6/ndisc.c	2006-03-20 06:53:29.000000000 +0100
+++ linux-2.6-patched/net/ipv6/ndisc.c	2006-03-22 11:47:03.000000000 +0100
@@ -1519,6 +1519,9 @@ static int ndisc_netdev_event(struct not
 		neigh_ifdown(&nd_tbl, dev);
 		fib6_run_gc(~0UL);
 		break;
+	case NETDEV_UNREGISTER:
+		fib6_run_gc(0);
+		break;
 	default:
 		break;
 	}
